### PR TITLE
Barometer glitch detection and correction

### PR DESCRIPTION
**Pull request by martinbudden**
Barometer glitch detection and correction.

### CODE REVIEW

**Comment by carlonb** [comment]
I saw the code mod, I'm asking why the glitch is assumed as 100 mt about, I think it's almost impossible a valid altitude change of 99mts in one baro evaluation loop.
The glitch value I think can be much lower than 100mt.
I hope to not misunderstood the code.

**Comment by digitalentity** [comment]
I would do it differently - maintain a mean value of barometric pressure and disregard everything that is outside a sane limit from the mean. But this will do as initial cut. I see, however, one problem - we don't let user know that barometer is glitching.

File: src/main/sensors/barometer.c [hunk 1]
```diff
@@ -196,13 +196,24 @@ void baroStartCalibration(void)
...
+*/
+#define PRESSURE_DELTA_GLITCH_THRESHOLD 1000
 static int32_t applyBarometerMedianFilter(int32_t newPressureReading)
 {
     static int32_t barometerFilterSamples[PRESSURE_SAMPLES_MEDIAN];
     static int currentFilterSampleIndex = 0;
     static bool medianFilterReady = false;
     int nextSampleIndex;
 
+    const int32_t previousPressureReading = barometerFilterSamples[currentFilterSampleIndex];
```
Comment (digitalentity) [inline src/main/sensors/barometer.c]: This is not a previous reading - we need to step back one item.

**Comment by martinbudden** [comment]
@carlonb , you have understood the code correctly. The glitch value probably should be less than 100m. However this is mainly to do with glitches caused by I2C errors to the barometer, which result in very large error readings, see this post: https://www.rcgroups.com/forums/showpost.php?p=37291228&postcount=13191

A more sophisticated approach can be implemented later.

**Comment by martinbudden** [comment]
@digitalentity , yes this is a somewhat "quick and dirty" implementation for an initial cut.

The way I'd like to eventually do it is to create a new filter type that does the median calculation and glitch detection. Once the filter code is separated from the barometer code it can be made more sophisticated and also have test code.

Agreed, it would be better if we could tell the user the barometer is glitching. Is there a suitable error reporting mechanism in iNav at the moment?

One other thing that we could do is record both the filtered and unfiltered baro readings in the blackbox log.

**Comment by carlonb** [comment]
@martinbudden ,
If intended as pure I2C error I agree with you, anyway this approach can also minimize other measure errors if the glitch detection value can be lower than 100mts.
Good idea to track baro glitch errors in bb log and via MSP.

**Review by digitalentity — APPROVED** [review]

**Pull request merged**